kvcoord: eagerly bail on Leaseholder not found in sendToReplicas

iskettaneh opened the pull request:
This commit introduces a couple of cluster settings:
1) eagerly_bail_on_leaseholder_not_found: If enabled, it will return
a replica unavailable error if we couldn't find the leaseholder in
sendToReplicas, and we expect that the range in unavailable.
2) leaseholder_not_found_backoff.threshold: Controls the threshold that
we should backoff when we can't find the leaseholder before returning
a replica unavailable error.

This is useful to prevent the dist sender from spinning for a long time
if the range is unavailable. This is especially important for leader
leases as we reject lease proposals if we don't know who the leader is,
which makes it unlikely that the per-replica circuit breaker would trip.

Fixes: #139638

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/140186)
<!-- Reviewable:end -->

iskettaneh committed:
kvcoord: eagerly bail on Leaseholder not found in sendToReplicas

This commit introduces a couple of cluster settings:
1) eagerly_bail_on_leaseholder_not_found: If enabled, it will return
a replica unavailable error if we couldn't find the leaseholder in
sendToReplicas, and we expect that the range is unavailable.
2) leaseholder_not_found_backoff.threshold: Controls the threshold that
we should backoff when we can't find the leaseholder before returning
a replica unavailable error.

This is useful to prevent the dist sender from spinning for a long time
if the range is unavailable. This is especially important for leader
leases as we reject lease proposals if we don't know who the leader is,
and it's the slow proposals that would trip the circuit breaker, which
makes it unlikely that the per-replica circuit breaker would trip with
leader leases.

Fixes: #139638

Release note: None

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140186)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2526 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIBIDSxDlRcRR8EwdNZ:-OIBIDSxDlRcRR8EwdN_:bqk8d3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2526)):*
> ```Go
> // isRangeUnavailableBasedOnNLHE determines whether the range is considered
> // unavailable based on the NLHEs seen during the execution of the batch.
> func (ds *DistSender) isRangeUnavailableBasedOnNLHE(
> ```

As a start, I wanted to do the basic scenario where all errors we get is NLHE/RPC. 
If we need to handle other cases more eagerly (without backing off), we could add those cases as we need them.


<!-- Sent from Reviewable.io -->

miraradeva reviewed (commented):
Looks great! I can't really vouch for the exact rule for returning the replica unavailable, but it sounds convincing. 

> Reviewed 7 of 7 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-:-OIWpT5ECPFuMgs6q8I0:baaxmu2)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @iskettaneh)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWk_Xo4WmDCZki5HNR:-OIWk_Xo4WmDCZki5HNS:b-se0fzh):*
s/in/is
___
*[`-- commits` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWkmNVAhsdgm94qOXg:-OIWkmNVAhsdgm94qOXh:bg6yg23):*
Maybe add that it's the slow/stuck lease proposals that trip the breakers?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 909 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWoCyG3Q_LrvudntrM:-OIWoCyG3Q_LrvudntrN:b-owg97i) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L909)):*
> ```Go
> 			// We expect a send error to be returned despite exceeding the backoff
> 			// threshold since there is other than NLHE.
> 			expectedErr: "sending to all replicas failed",
> ```

Don't we expect the range-not-found error to be returned? Or is that included in the sending error?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlXuN3RY8Ws-WOpvN:-OIWlXuN3RY8Ws-WOpvO:bzdqhqh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L419)):*
> ```Go
> 
> var EagerlyBailOnLeaseholderNotFound = settings.RegisterBoolSetting(
> 	settings.ApplicationLevel,
> ```

Is there a reason these are `ApplicationLevel` and not `SystemOnly`?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 422 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWl7ED7xL_em5rl3i6:-OIWl7ED7xL_em5rl3i7:b-ekhwr5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L422)):*
> ```Go
> 	"kv.dist_sender.eagerly_bail_on_leaseholder_not_found.enabled",
> 	"when true, the dist sender will eagerly return an error when it encounters errors that "+
> 		"indicates an unavailable range, rather than retrying until the circuit breaker trips",
> ```

s/indicates/indicate
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlEwN52vq9Y0W0PRE:-OIWlEwN52vq9Y0W0PRF:b-xfoqis) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L423)):*
> ```Go
> 	"when true, the dist sender will eagerly return an error when it encounters errors that "+
> 		"indicates an unavailable range, rather than retrying until the circuit breaker trips",
> 	true,
> ```

Should this always be true by default? Maybe only after leader leases are on by default? Or metamorphically on for now?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlkAD-TedUdnX_87k:-OIWlkAD-TedUdnX_87l:b88j12w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L432)):*
> ```Go
> 		"returning that the range in unavailable. This only works when "+
> 		"eagerly_bail_on_leaseholder_not_found is set to true",
> 	time.Second*6,
> ```

Should this match the duration we wait before tripping circuit breakers?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2592 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWowKRCM0lvP_dSkw4:-OIWowKRCM0lvP_dSkw5:b8tbezn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2592)):*
> ```Go
> 	routing rangecache.EvictionToken,
> 	withCommit bool,
> 	curDuration time.Duration,
> ```

nit: maybe rename to something more descriptive? Like `timeSinceSendStart` or something more concise. 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2714 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWmrWXDQZva54hyN8Q:-OIWmrWXDQZva54hyN8R:be8aha1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2714)):*
> ```Go
> 				seenNLHE = true
> 			} else {
> 				seenOtherThanNLHE = true
> ```

Do RPC errors not fit in either of these cases? Probably worth a comment.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2734 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWnFxF7jyt8z8r0cBp:-OIWnFxF7jyt8z8r0cBq:b-win9io) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2734)):*
> ```Go
> 						replDesc = *routing.Leaseholder()
> 					} else {
> 						replDesc = prevReplica
> ```

nit: Maybe add a comment for this case too? Why do we include `prevReplica`?


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140186)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @miraradeva)

___
*[`-- commits` line 7 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWk_Xo4WmDCZki5HNR:-OImB7IEDTD67BljANZv:b-896fix):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

s/in/is
</blockquote></details>

Done.
___
*[`-- commits` line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWkmNVAhsdgm94qOXg:-OImBPSY7a0jQt7e9ypC:b-896fix):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Maybe add that it's the slow/stuck lease proposals that trip the breakers?
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlXuN3RY8Ws-WOpvN:-OImByDz7xpgVugs_rvF:boixuq6) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L419)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Is there a reason these are `ApplicationLevel` and not `SystemOnly`?
</blockquote></details>

Not a specific reason, but I noticed that the other cluster settings in the DistSender are all application level. This got me thinking that since the DistSender is actually the KV client (and not the kvserver), maybe it's fine to make it application level.
That being said, I am find with changing it to SystemOnly if it's better in our case.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 422 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWl7ED7xL_em5rl3i6:-OImCV2N54II5aFigN0a:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L422)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

s/indicates/indicate
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlEwN52vq9Y0W0PRE:-OImCYe60Bt_7TKiFVkW:bsfuwog) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L423)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Should this always be true by default? Maybe only after leader leases are on by default? Or metamorphically on for now?
</blockquote></details>

That's a good point. However, I thought that we could get more millage on this feature if we enable it from now.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlkAD-TedUdnX_87k:-OImCkpM1K78RcmjelPR:b2lxtu1) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L432)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Should this match the duration we wait before tripping circuit breakers?
</blockquote></details>

The circuit breaker gets triggered if the Raft replication is being slow (60 seconds by default).
However, if we really think that the range in unavailable, maybe we should really wait the full 60 seconds.

Note that the circuit breaker is still functional with leader leases. It's just that now it doesn't trip on unavailable ranges.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2714 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWmrWXDQZva54hyN8Q:-OImErnCA-0MpLkuuzwE:boc2hki) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2714)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Do RPC errors not fit in either of these cases? Probably worth a comment.
</blockquote></details>

I have a comment above this if-statement that says that it's non RPC error. But I guess that might still be confusing so I added an inline comment to mention that more clearly.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2734 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWnFxF7jyt8z8r0cBp:-OImDXUM0ymV2QUh3RXr:bg058ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2734)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

nit: Maybe add a comment for this case too? Why do we include `prevReplica`?
</blockquote></details>

That's a good point actually. Right now, the `prevReplica` is selected arbitrarily if we don't have any leaseholder in our transport (which is unlikely, but possible).
@arulajmani Do you have a preference on which replica should we return in this case?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 909 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWoCyG3Q_LrvudntrM:-OImDqi_1mctUqaj0EzE:b-b4lvxa) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L909)):*
<details><summary><i>Previously, miraradeva (Mira Radeva) wrote…</i></summary><blockquote>

Don't we expect the range-not-found error to be returned? Or is that included in the sending error?
</blockquote></details>

Yes, the send error is its own path which selects the best error that it should return. That part hasn't been touched in this PR.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
> Reviewed 1 of 7 files at r1, 3 of 5 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140186)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh and @miraradeva)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlXuN3RY8Ws-WOpvN:-OJTmNUZD_4d6xZrxqch:bp3dmv2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L419)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Not a specific reason, but I noticed that the other cluster settings in the DistSender are all application level. This got me thinking that since the DistSender is actually the KV client (and not the kvserver), maybe it's fine to make it application level.
That being said, I am find with changing it to SystemOnly if it's better in our case.
</blockquote></details>

I think `ApplicationLevel` makes sense here, as SQL pods can alter this setting if they want.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 423 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlEwN52vq9Y0W0PRE:-OJTeyP3Db5_OGpm9vsQ:b-i46cq4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L423)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

That's a good point. However, I thought that we could get more millage on this feature if we enable it from now.
</blockquote></details>

Maybe we should enable leader leases on master now that 25.1 is out as well, which will make the discussion here moot.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlkAD-TedUdnX_87k:-OJTf6673o_Wlf8G4Rl_:b5v0226) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L432)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

The circuit breaker gets triggered if the Raft replication is being slow (60 seconds by default).
However, if we really think that the range in unavailable, maybe we should really wait the full 60 seconds.

Note that the circuit breaker is still functional with leader leases. It's just that now it doesn't trip on unavailable ranges.
</blockquote></details>

Should we change the default to be something slightly higher than a lease interval, just to be safe -- maybe 10s? 
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2734 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWnFxF7jyt8z8r0cBp:-OJTiKGa8rgosTupsnf4:bqyk03w) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L2734)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

That's a good point actually. Right now, the `prevReplica` is selected arbitrarily if we don't have any leaseholder in our transport (which is unlikely, but possible).
@arulajmani Do you have a preference on which replica should we return in this case?
</blockquote></details>

Not really; maybe we encode whether the replica was arbitrarily chosen by the DistSender or not in the "cause"? That should help differentiate these two cases when debugging.

----

Coming back to this, this seems similar to how we handle ReplicaUnavailableErrors below.
___
*[`pkg/kv/kvclient/kvcoord/send_test.go` line 262 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJTmDgY79UAleAWuAu1:-OJTmDgY79UAleAWuAu2:b139llb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/send_test.go#L262)):*
> ```Go
> 	routing, err := ds.getRoutingInfo(ctx, desc.StartKey, rangecache.EvictionToken{}, false /* useReverseScan */)
> 	require.NoError(t, err)
> 	return ds.sendToReplicas(ctx, &kvpb.BatchRequest{}, routing, false /* withCommit */, time.Millisecond /* curDuration */)
> ```

nit: instead of a millisecond, let's pass in 0 value duration,
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2731 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJThq7HBUvsCK3TR_IP:-OJThq7HBUvsCK3TR_IQ:bnkz5ud) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender.go#L2731)):*
> ```Go
> 			// request for a long time. However, we first need to check if we should
> 			// backoff before returning the error.
> 			if !ds.shouldBackoffRangeUnavailableBasedOnNLHE(durationSinceStart) {
> ```

I may be missing something, but where are we backing off?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2741 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJTi09wAi9FOzDjGfb9:-OJTi09wAi9FOzDjGfbA:by5r5b1) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender.go#L2741)):*
> ```Go
> 					}
> 					replicaUnavailableError = kvpb.NewReplicaUnavailableError(
> 						errors.New("Leaseholder is unavailable"), desc, replDesc)
> ```

nit: error strings should begin with lower case letter.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 3129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJTl73G0_l_O2jNQde9:-OJTl73G0_l_O2jNQdeA:bcm6px5) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender.go#L3129)):*
> ```Go
> 						leaseholderUnavailable = false
> 						routeToLeaseholder = true
> 						seenNLHE = false
> ```

This seems weird -- why're we setting `seenNLHE` false in the case where we're handling an `NLHE`?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJTlxP52I-FeApRHdHK:-OJTlxP52I-FeApRHdHL:b-sojrie) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L52)):*
> ```Go
> 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
> 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
> 	"github.com/cockroachdb/crlib/crtime"
> ```

Why crtime? Should we use `time.Now` instead?


<!-- Sent from Reviewable.io -->

iskettaneh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140186)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @arulajmani and @miraradeva)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 419 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlXuN3RY8Ws-WOpvN:-OJYqGLh6425M8ikeWZW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L419)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I think `ApplicationLevel` makes sense here, as SQL pods can alter this setting if they want.
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 432 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OIWlkAD-TedUdnX_87k:-OJYqPu91c7mj_8d_5XL:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0134373859c29aeaf7cc19be529f3d6ce498a1a/pkg/kv/kvclient/kvcoord/dist_sender.go#L432)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Should we change the default to be something slightly higher than a lease interval, just to be safe -- maybe 10s?
</blockquote></details>

Done.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2731 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJThq7HBUvsCK3TR_IP:-OJYqSw88xJ_HQgj6U3f:bqoxfz) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender.go#L2731)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

I may be missing something, but where are we backing off?
</blockquote></details>

The caller to this function does the backoff. I added a comment to make it more obvious.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 3129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJTl73G0_l_O2jNQde9:-OJYrQej05MrcAt9iGGt:b-ijs491) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender.go#L3129)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

This seems weird -- why're we setting `seenNLHE` false in the case where we're handling an `NLHE`?
</blockquote></details>

Since this codepath resets the transport and retries all replicas again, I thought it would be nice to reset those booleans since they will be set as needed on the next transport iterations.
I added a comment to clarify this part.
___
*[`pkg/kv/kvclient/kvcoord/dist_sender_test.go` line 52 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJTlxP52I-FeApRHdHK:-OJYsAbLAZKxsNaJlVFr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender_test.go#L52)):*
<details><summary><i>Previously, arulajmani (Arul Ajmani) wrote…</i></summary><blockquote>

Why crtime? Should we use `time.Now` instead?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

arulajmani reviewed (commented):
Sorry for letting this sit. Should be good to go once this comment is addressed.

> Reviewed 1 of 7 files at r1, 5 of 6 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-:-OK7mrgxEpJAdkey4vPf:b-sw6h2t)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @iskettaneh and @miraradeva)

___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 2731 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJThq7HBUvsCK3TR_IP:-OK7lW7CD82EQYWeTqXo:b-it15tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender.go#L2731)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

The caller to this function does the backoff. I added a comment to make it more obvious.
</blockquote></details>

I see. In that case, I think it makes more sense for the caller to call this function and decide whether it wants to backoff or not. Here, we can simply set `replicaUnavailableError` blindly.

Does this also mean we no longer nead to thread in `durationSinceStart` to `sendToReplicas`?
___
*[`pkg/kv/kvclient/kvcoord/dist_sender.go` line 3129 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OJTl73G0_l_O2jNQde9:-OK7lMN2-DZD1X4be3n0:b-i4mx8j) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca937ab95d21030a8603595a38965c7c24f89547/pkg/kv/kvclient/kvcoord/dist_sender.go#L3129)):*
<details><summary><i>Previously, iskettaneh wrote…</i></summary><blockquote>

Since this codepath resets the transport and retries all replicas again, I thought it would be nice to reset those booleans since they will be set as needed on the next transport iterations.
I added a comment to clarify this part.
</blockquote></details>

I see, this makes sense. Is it worth abstracting this into a function? Maybe `resetTransportAndStartAfresh` or something?
___
*[`.cursorignore` line 1 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/140186#-OK7iA4Y7j9bAzGTo9Md:-OK7iA4ZENWF0MEtsZmY:bdh83ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/4678d5422ff696e0686e875992cda2e3b271ad41/.cursorignore#L1)):*
> ```
> # ignore everything
> ```

I assume you don't wanna check this in?


<!-- Sent from Reviewable.io -->
